pkg/zcapld/resolve.go

```diff
@@ -51,34 +53,48 @@ func (s SimpleKeyResolver) Resolve(keyID string) (*verifier.PublicKey, error) {
 	return key, nil
 }
 
-// DIDKeyResolver resolves verification keys from did:key URLs: https://w3c-ccg.github.io/did-method-key/.
-type DIDKeyResolver struct {
+type kmsProvider struct{ kms.KeyManager }
+
+func (p *kmsProvider) KMS() kms.KeyManager {
+	return p
+}
+
+// NewDIDResolver creates new DID resolver.
+func NewDIDResolver(k kms.KeyManager, v VDRResolver) *DIDResolver {
+	if v != nil {
+		return &DIDResolver{KMS: k, VDR: v}
+	}
+
+	return &DIDResolver{KMS: k, VDR: vdr.New(&kmsProvider{KeyManager: k}, vdr.WithVDR(vdrkey.New()))}
+}
+
+// DIDResolver resolves verification keys from did:key URLs: https://w3c-ccg.github.io/did-method-key/.
+type DIDResolver struct {
+	KMS kms.KeyManager
+	VDR VDRResolver
 }
 
 // Resolve expects 'didKeyURL' to be a did:key URL.
 // Example: "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH#z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH".
-func (d *DIDKeyResolver) Resolve(didKeyURL string) (*verifier.PublicKey, error) {
-	const numParts = 2
+func (d *DIDResolver) Resolve(id string) (*verifier.PublicKey, error) {
+	didToResolve := id
 
-	parts := strings.Split(didKeyURL, "#")
-	if len(parts) != numParts {
-		return nil, fmt.Errorf("not a did:key URL: %s", didKeyURL)
+	if strings.HasPrefix(id, "did:key") {
+		didToResolve = strings.Split(id, "#")[0]
 	}
 
-	docResolution, err := didkey.New().Read(parts[0])
+	docResolution, err := d.VDR.Resolve(didToResolve)
 	if err != nil {
-		return nil, fmt.Errorf("failed to parse url %s: %w", parts[0], err)
+		return nil, fmt.Errorf("failed to parse url %s: %w", didToResolve, err)
 	}
 
 	for _, vm := range docResolution.DIDDocument.VerificationMethods(did.CapabilityDelegation)[did.CapabilityDelegation] {
-		if parts[1] == vm.VerificationMethod.ID || didKeyURL == vm.VerificationMethod.ID {
-			return &verifier.PublicKey{
-				Type:  vm.VerificationMethod.Type,
-				Value: vm.VerificationMethod.Value,
-				JWK:   vm.VerificationMethod.JSONWebKey(),
-			}, nil
-		}
+		return &verifier.PublicKey{
```

Comment:
Must not delete the check that matches the verification method's ID. A DID document can have several verification methods and the DID URL is pointing to exactly one of those.